py/server/tests/test_partitioned_table_service.py

```diff
@@ -0,0 +1,248 @@
+#
+# Copyright (c) 2016-2024 Deephaven Data Labs and Patent Pending
+#
+
+import threading
+import time
+import unittest
+from typing import Callable, Tuple, Optional, Generator, List
+
+import numpy as np
+import pyarrow as pa
+import pyarrow.compute as pc
+
+from deephaven import new_table, garbage_collect
+from deephaven.column import byte_col, char_col, short_col, int_col, long_col, float_col, double_col, string_col, \
+    datetime_col, bool_col, ColumnType
+from deephaven.execution_context import get_exec_ctx, ExecutionContext
+from deephaven.experimental.partitioned_table_service import PartitionedTableServiceBackend, TableKey, \
+    PartitionedTableLocationKey, PythonTableDataService
+import deephaven.arrow as dharrow
+from deephaven.liveness_scope import liveness_scope
+
+from tests.testbase import BaseTestCase
+
+
+class TestBackend(PartitionedTableServiceBackend):
+    def __init__(self, gen_pa_table: Generator[pa.Table, None, None], pt_schema: pa.Schema,  pc_schema: Optional[pa.Schema] = None):
+        self.pt_schema = pt_schema
+        self.pc_schema = pc_schema
+        self.gen_pa_table = gen_pa_table
+        self.sub_new_partition_cancelled = False
+        self.partitions: dict[PartitionedTableLocationKey, pa.Table] = {}
+        self.partitions_size_subscriptions: dict[PartitionedTableLocationKey, bool] = {}
+        self.existing_partitions_called = 0
+        self.partition_size_called = 0
+
+    def table_schema(self, table_key: TableKey) -> Tuple[pa.Schema, Optional[pa.Schema]]:
+        if table_key.key == "test":
+             return self.pt_schema, self.pc_schema
+        return pa.Schema(), None
+
+    def existing_partitions(self, table_key: TableKey, callback: Callable[[PartitionedTableLocationKey, Optional[pa.Table]], None]) -> None:
+        pa_table = next(self.gen_pa_table)
+        if table_key.key == "test":
+            ticker = str(pa_table.column("Ticker")[0])
+
+            partition_key = PartitionedTableLocationKey(f"{ticker}/NYSE")
+            self.partitions[partition_key] = pa_table
+
+            expr = ((pc.field("Ticker") == f"{ticker}") & (pc.field("Exchange") == "NYSE"))
+            callback(partition_key, pa_table.filter(expr).select(["Ticker", "Exchange"]).slice(0, 1))
+            self.existing_partitions_called += 1
+
+    def partition_size(self, table_key: TableKey, table_location_key: PartitionedTableLocationKey,
+                       callback: Callable[[int], None]) -> None:
```

Comment:
Since I'm having to try to decrypt what is going on, should `table_location_key` be called `partition_key`?  I think it is the key for the partition within a table.

Reply:
No. Firstly, the underlying infrastructure this maps to use the term `TableLocationKey`. Secondly, a TLK may be part of multiple partitions (denoted by partitioning column values), and multiple TLKs may be part of the same set of partitions (meaning, same partitioning column values) but still be distinct locations, e.g. multiple `.parquet` files in the same directory. (Side note: In a case of convergent evolution, Iceberg seems to have settled on almost the same terminology many years ago.)